File: backend/go/internal/explore/message_content.go
package explore

import (
	"encoding/json"
	"math"
	"strconv"
	"strings"
	"time"
)

const richMessagePrefix = "[[MRMSG]]"

type outboundVoiceMessagePayload struct {
	DurationSec int       `json:"durationSec"`
	Kind        string    `json:"kind"`
	MimeType    string    `json:"mimeType"`
	SizeBytes   int64     `json:"sizeBytes"`
	Title       string    `json:"title"`
	VoiceID     string    `json:"voiceId"`
	VoiceURL    string    `json:"voiceUrl"`
	Waveform    []float64 `json:"waveform"`
}

type outboundPhotoMessagePayload struct {
	Kind      string `json:"kind"`
	MimeType  string `json:"mimeType"`
	SizeBytes int64  `json:"sizeBytes"`
	Title     string `json:"title"`
	URL       string `json:"url"`
}

type outboundLocationMessagePayload struct {
	Kind          string   `json:"kind"`
	Latitude      *float64 `json:"latitude"`
	LocationLabel string   `json:"locationLabel"`
	Longitude     *float64 `json:"longitude"`
	Title         string   `json:"title"`
}

type richMessageEnvelope struct {
	Kind string `json:"kind"`
}

type parsedMessageContent struct {
	kind            MessageContentKind
	locationMessage *LocationMessageAsset
	photoMessage    *PhotoMessageAsset
	preview         string
	voiceMessage    *VoiceMessageAsset
}

func clamp01(value float64) float64 {
	if math.IsNaN(value) || math.IsInf(value, 0) {
		return 0
	}
	return math.Min(1, math.Max(0, value))
}

func normalizeWaveform(values []float64) []float64 {
	if len(values) == 0 {
		return nil
	}

	normalized := make([]float64, 0, min(len(values), 256))
	for _, value := range values {
		next := clamp01(value)
		if next <= 0 {
			continue
		}
		normalized = append(normalized, next)
		if len(normalized) >= 256 {
			break
		}
	}
	if len(normalized) == 0 {
		return nil
	}
	return normalized
}

func normalizeVoiceDuration(value int, fallback int) int {
	if value <= 0 {
		return fallback
	}
	return value
}

func normalizePreview(preview string, fallback string) string {
	trimmed := strings.TrimSpace(preview)
	if trimmed != "" {
		return trimmed
	}
	return strings.TrimSpace(fallback)
}

func buildTextContent(body string, preview string) parsedMessageContent {
	trimmedBody := strings.TrimSpace(body)
	return parsedMessageContent{
		kind:    MessageContentKindText,
		preview: normalizePreview(preview, firstNonEmpty(trimmedBody, "Yeni mesaj")),
	}
}

func buildVoiceContent(payload outboundVoiceMessagePayload, preview string) parsedMessageContent {
	durationSec := normalizeVoiceDuration(payload.DurationSec, 6)
	return parsedMessageContent{
		kind:    MessageContentKindVoice,
		preview: normalizePreview(preview, "Sesli mesaj ("+itoa(durationSec)+" sn)"),
		voiceMessage: &VoiceMessageAsset{
			ConversationID: "",
			CreatedAt:      time.Time{},
			DurationSec:    durationSec,
			FileName:       "",
			ID:             strings.TrimSpace(payload.VoiceID),
			MimeType:       strings.TrimSpace(payload.MimeType),
			SizeBytes:      maxInt64(0, payload.SizeBytes),
			URL:            strings.TrimSpace(payload.VoiceURL),
			Waveform:       normalizeWaveform(payload.Waveform),
		},
	}
}

func buildPhotoContent(payload outboundPhotoMessagePayload, preview string) parsedMessageContent {
	title := firstNonEmpty(strings.TrimSpace(payload.Title), "Fotograf")
	return parsedMessageContent{
		kind:    MessageContentKindPhoto,
		preview: normalizePreview(preview, "Fotograf: "+title),
		photoMessage: &PhotoMessageAsset{
			MimeType:  strings.TrimSpace(payload.MimeType),
			SizeBytes: maxInt64(0, payload.SizeBytes),
			Title:     title,
			URL:       strings.TrimSpace(payload.URL),
		},
	}
}

func buildLocationContent(payload outboundLocationMessagePayload, preview string) parsedMessageContent {
	locationLabel := firstNonEmpty(strings.TrimSpace(payload.LocationLabel), "Konum")
	return parsedMessageContent{
		kind:    MessageContentKindLocation,
		preview: normalizePreview(preview, "Konum: "+locationLabel),
		locationMessage: &LocationMessageAsset{
			Latitude:      payload.Latitude,
			LocationLabel: locationLabel,
			Longitude:     payload.Longitude,
			Title:         firstNonEmpty(strings.TrimSpace(payload.Title), "Konum"),
		},
	}
}

func ParseMessageContent(
	body string,
	fallbackKind MessageContentKind,
	fallbackPreview string,
	fallbackVoice *VoiceMessageAsset,
	fallbackPhoto *PhotoMessageAsset,
	fallbackLocation *LocationMessageAsset,
) parsedMessageContent {
	if fallbackKind == MessageContentKindVoice && fallbackVoice != nil {
		voiceCopy := *fallbackVoice
		voiceCopy.Waveform = normalizeWaveform(voiceCopy.Waveform)
		return parsedMessageContent{
			kind:         MessageContentKindVoice,
			preview:      normalizePreview(fallbackPreview, "Sesli mesaj ("+itoa(normalizeVoiceDuration(voiceCopy.DurationSec, 6))+" sn)"),
			voiceMessage: &voiceCopy,
		}
	}
	if fallbackKind == MessageContentKindPhoto && fallbackPhoto != nil {
		photoCopy := *fallbackPhoto
		return parsedMessageContent{
			kind:         MessageContentKindPhoto,
			photoMessage: &photoCopy,
			preview:      normalizePreview(fallbackPreview, "Fotograf"),
		}
	}
	if fallbackKind == MessageContentKindLocation && fallbackLocation != nil {
		locationCopy := *fallbackLocation
		locationLabel := firstNonEmpty(strings.TrimSpace(locationCopy.LocationLabel), "Konum")
		locationCopy.LocationLabel = locationLabel
		locationCopy.Title = firstNonEmpty(strings.TrimSpace(locationCopy.Title), "Konum")
		return parsedMessageContent{
			kind:            MessageContentKindLocation,
			locationMessage: &locationCopy,
			preview:         normalizePreview(fallbackPreview, "Konum: "+locationLabel),
		}
	}

	raw := body
	if !strings.HasPrefix(raw, richMessagePrefix) {
		return buildTextContent(raw, fallbackPreview)
	}

	encoded := strings.TrimPrefix(raw, richMessagePrefix)
	var envelope richMessageEnvelope
	if err := json.Unmarshal([]byte(encoded), &envelope); err != nil {
		return buildTextContent(raw, fallbackPreview)
	}

	switch strings.ToLower(strings.TrimSpace(envelope.Kind)) {
	case "voice":
		var payload outboundVoiceMessagePayload
		if err := json.Unmarshal([]byte(encoded), &payload); err != nil {
			return buildTextContent(raw, fallbackPreview)
		}
		return buildVoiceContent(payload, fallbackPreview)
	case "photo":
		var payload outboundPhotoMessagePayload
		if err := json.Unmarshal([]byte(encoded), &payload); err != nil {
			return buildTextContent(raw, fallbackPreview)
		}
		return buildPhotoContent(payload, fallbackPreview)
	case "location":
		var payload outboundLocationMessagePayload
		if err := json.Unmarshal([]byte(encoded), &payload); err != nil {
			return buildTextContent(raw, fallbackPreview)
		}
		return buildLocationContent(payload, fallbackPreview)
	default:
		return buildTextContent(raw, fallbackPreview)
	}
}

func EncodeVoiceMessageBody(asset VoiceMessageAsset) (string, error) {
	payload := outboundVoiceMessagePayload{
		DurationSec: normalizeVoiceDuration(asset.DurationSec, 6),
		Kind:        "voice",
		MimeType:    strings.TrimSpace(asset.MimeType),
		SizeBytes:   maxInt64(0, asset.SizeBytes),
		Title:       "Sesli mesaj",
		VoiceID:     strings.TrimSpace(asset.ID),
		VoiceURL:    strings.TrimSpace(asset.URL),
		Waveform:    normalizeWaveform(asset.Waveform),
	}
	raw, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}
	return richMessagePrefix + string(raw), nil
}

func HydrateConversationMessage(message ConversationMessage) ConversationMessage {
	content := ParseMessageContent(
		message.Body,
		message.Kind,
		message.Preview,
		message.VoiceMessage,
		message.PhotoMessage,
		message.LocationMessage,
	)
	message.Kind = content.kind
	message.LocationMessage = content.locationMessage
	message.PhotoMessage = content.photoMessage
	message.Preview = content.preview
	message.VoiceMessage = content.voiceMessage
	return message
}

func HydrateConversationSummary(summary ConversationSummary) ConversationSummary {
	content := ParseMessageContent(
		summary.LastMessage,
		summary.LastMessageKind,
		summary.LastMessagePreview,
		summary.LastVoiceMessage,
		summary.LastPhotoMessage,
		summary.LastLocationMessage,
	)
	summary.LastLocationMessage = content.locationMessage
	summary.LastMessageKind = content.kind
	summary.LastMessagePreview = content.preview
	summary.LastPhotoMessage = content.photoMessage
	summary.LastVoiceMessage = content.voiceMessage
	return summary
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return strings.TrimSpace(value)
		}
	}
	return ""
}

func itoa(value int) string {
	return strconv.Itoa(value)
}

func maxInt64(a int64, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}

File: backend/node/lib/message-content.js
const { normalizeText } = require('./utils');

const RICH_MESSAGE_PREFIX = '[[MRMSG]]';

function clamp01(value) {
  if (!Number.isFinite(value)) {
    return 0;
  }
  return Math.min(1, Math.max(0, value));
}

function normalizeWaveform(values) {
  if (!Array.isArray(values)) {
    return [];
  }

  return values
    .map(value => Number(value))
    .filter(value => Number.isFinite(value))
    .map(value => clamp01(value))
    .filter(value => value > 0)
    .slice(0, 256);
}

function normalizeVoiceDuration(value, fallback = 6) {
  if (!Number.isFinite(value)) {
    return fallback;
  }
  return Math.max(1, Math.floor(Number(value)));
}

function buildTextContent(body, preview) {
  const fallbackPreview = normalizeText(body) || 'Yeni mesaj';
  return {
    kind: 'text',
    locationMessage: null,
    photoMessage: null,
    preview: normalizeText(preview, fallbackPreview),
    voiceMessage: null,
  };
}

function buildVoiceContent(payload, preview) {
  const durationSec = normalizeVoiceDuration(payload?.durationSec, 6);
  return {
    kind: 'voice',
    locationMessage: null,
    photoMessage: null,
    preview: normalizeText(preview, `Sesli mesaj (${durationSec} sn)`),
    voiceMessage: {
      conversationId: '',
      createdAt: '',
      durationSec,
      fileName: '',
      id: normalizeText(payload?.voiceId),
      mimeType: normalizeText(payload?.mimeType),
      sizeBytes: Number.isFinite(payload?.sizeBytes)
        ? Math.max(0, Number(payload.sizeBytes))
        : 0,
      url: normalizeText(payload?.voiceUrl),
      waveform: normalizeWaveform(payload?.waveform),
    },
  };
}

function buildPhotoContent(payload, preview) {
  const title = normalizeText(payload?.title, 'Fotograf');
  const resolvedPreview = normalizeText(preview, `Fotograf: ${title}`);
  return {
    kind: 'photo',
    locationMessage: null,
    photoMessage: {
      mimeType: normalizeText(payload?.mimeType),
      sizeBytes: Number.isFinite(payload?.sizeBytes)
        ? Math.max(0, Number(payload.sizeBytes))
        : 0,
      title,
      url: normalizeText(payload?.url),
    },
    preview: resolvedPreview,
    voiceMessage: null,
  };
}

function buildLocationContent(payload, preview) {
  const locationLabel = normalizeText(payload?.locationLabel, 'Konum');
  return {
    kind: 'location',
    locationMessage: {
      latitude: Number.isFinite(payload?.latitude) ? Number(payload.latitude) : undefined,
      locationLabel,
      longitude: Number.isFinite(payload?.longitude) ? Number(payload.longitude) : undefined,
      title: normalizeText(payload?.title, 'Konum'),
    },
    photoMessage: null,
    preview: normalizeText(preview, `Konum: ${locationLabel}`),
    voiceMessage: null,
  };
}

function parseMessageContent(body, fallback = {}) {
  const raw = typeof body === 'string' ? body : '';

  if (fallback.kind === 'voice' && fallback.voiceMessage) {
    const durationSec = normalizeVoiceDuration(fallback.voiceMessage.durationSec, 6);
    return {
      kind: 'voice',
      locationMessage: null,
      photoMessage: null,
      preview: normalizeText(fallback.preview, `Sesli mesaj (${durationSec} sn)`),
      voiceMessage: {
        ...fallback.voiceMessage,
        waveform: normalizeWaveform(fallback.voiceMessage.waveform),
      },
    };
  }

  if (fallback.kind === 'photo' && fallback.photoMessage) {
    return {
      kind: 'photo',
      locationMessage: null,
      photoMessage: fallback.photoMessage,
      preview: normalizeText(fallback.preview, 'Fotograf'),
      voiceMessage: null,
    };
  }

  if (fallback.kind === 'location' && fallback.locationMessage) {
    const locationLabel = normalizeText(
      fallback.locationMessage.locationLabel,
      'Konum',
    );
    return {
      kind: 'location',
      locationMessage: fallback.locationMessage,
      photoMessage: null,
      preview: normalizeText(fallback.preview, `Konum: ${locationLabel}`),
      voiceMessage: null,
    };
  }

  if (!raw.startsWith(RICH_MESSAGE_PREFIX)) {
    return buildTextContent(raw, fallback.preview);
  }

  try {
    const parsed = JSON.parse(raw.slice(RICH_MESSAGE_PREFIX.length));
    if (!parsed || typeof parsed !== 'object' || typeof parsed.kind !== 'string') {
      return buildTextContent(raw, fallback.preview);
    }

    if (parsed.kind === 'voice') {
      return buildVoiceContent(parsed, fallback.preview);
    }
    if (parsed.kind === 'photo') {
      return buildPhotoContent(parsed, fallback.preview);
    }
    if (parsed.kind === 'location') {
      return buildLocationContent(parsed, fallback.preview);
    }
  } catch {
    return buildTextContent(raw, fallback.preview);
  }

  return buildTextContent(raw, fallback.preview);
}

function enrichConversationMessage(message) {
  const content = parseMessageContent(message.body, {
    kind: message.kind,
    locationMessage: message.locationMessage,
    photoMessage: message.photoMessage,
    preview: message.preview,
    voiceMessage: message.voiceMessage,
  });

  return {
    ...message,
    kind: content.kind,
    locationMessage: content.locationMessage,
    photoMessage: content.photoMessage,
    preview: content.preview,
    voiceMessage: content.voiceMessage,
  };
}

function enrichConversationSummary(summary) {
  const content = parseMessageContent(summary.lastMessage, {
    kind: summary.lastMessageKind,
    locationMessage: summary.lastLocationMessage,
    photoMessage: summary.lastPhotoMessage,
    preview: summary.lastMessagePreview,
    voiceMessage: summary.lastVoiceMessage,
  });

  return {
    ...summary,
    lastLocationMessage: content.locationMessage,
    lastMessageKind: content.kind,
    lastMessagePreview: content.preview,
    lastPhotoMessage: content.photoMessage,
    lastVoiceMessage: content.voiceMessage,
  };
}

module.exports = {
  RICH_MESSAGE_PREFIX,
  enrichConversationMessage,
  enrichConversationSummary,
  parseMessageContent,
};
